include/aspect/utilities.h

```diff
@@ -1064,6 +1065,47 @@ namespace aspect
          */
         const std::function<Tensor<1,dim> (const Point<dim> &)> function_object;
     };
+
+    /**
+     * Create a permutation vector which can be used by the apply_permutation function
+     * which sorts the vector in increasing order.
+     *
+     * @param vector vector to sort
+     */
+    template <typename T>
+    std::vector<std::size_t>
+    compute_sorting_permutation(const std::vector<T> &vector);
+
+    /**
+     * Applies a permutation vector to another vector.
+     *
+     * @param vector vector to sort
+     * @param permutation_vector The permutation vector used to sort the input vector.
+     * @return std::vector<T>
+     */
+    template <typename T>
+    std::vector<T>
+    apply_permutation(
+      const std::vector<T> &vec,
+      const std::vector<std::size_t> &permutation_vector);
+
+    /**
+     * Get volume weighted rotation matrices, using random draws to convert
+     * to a discrete number of orientations, weighted by volume.
+     * The input is a vector of volume fractions and a vector of rotation matrices.
+     * The vectors need to have the same length.
+     *
+     * @param volume_fraction a vector of doubles representing the volume fraction of each grain
+     * @param rotation_matrices a vector of 2nd order 3D tensors representing the rotation matrix of each grain
+     * @param n_output_grains The number of rotation matrices which are output by this function. This can be
+     * different from the number of entries in the volume fraction and rotation matrices vectors.
+     * @param random_number_generator a reference to a mt19937 random number generator.
+     */
+    std::vector<Tensor<2,3>>
+    random_draw_volume_weighting_rotation_matrices(const std::vector<double> volume_fraction,
+                                                   const std::vector<Tensor<2,3>> rotation_matrices,
+                                                   const unsigned int n_output_grains,
```

Comment:
This name doesn't make sense to me.

Reply:
I changed the name to `rotation_matrices_random_draw_volume_weighting`, I hope that is slightly better. I am open for suggestions though.

Reply:
Sorry, I should have been more precise: I was talking about the int argument.

Reply:
ah, oke. I chaned it to `n_output_matrices`